camptix.php

```diff
@@ -2810,24 +2811,24 @@ function generate_attendee_report( $format ) {
 			 */
 			$this->filter_post_meta = $this->prepare_metadata_for( $attendee_ids );
 			unset( $attendee_ids, $attendee );
+			$buyer_map = array();
+
 
 			foreach ( $attendees as $attendee ) {
 				$attendee_id = $attendee->ID;
 
-				$buyer = get_posts( array(
-					'post_type'      => 'tix_attendee',
-					'post_status'    => array( 'publish', 'pending' ),
-					'posts_per_page' => 1,
-					'orderby'        => 'ID',
-					'order'          => 'ASC',
+				$access_token = get_post_meta( $attendee->ID, 'tix_access_token', true );
 
-					'meta_query'     => array(
-						array(
-							'key'    => 'tix_access_token',
-							'value'  => get_post_meta( $attendee->ID, 'tix_access_token', true ),
-						),
-					),
-				) );
+				/**
+				 So when a buyer buys tickets for a bunch of attendees, access token for all those attendees would be same.
+				 Further, the buyer will always be inserted first into the database.
+				 Which means that when access token for a lot attendees is same, we can figure out the buyer by finding the first attendee with same access token.
+				 * **/
+				if ( ! isset( $buyer_map[ $access_token ] ) ) {
+					$buyer_map[ $access_token ] = $attendee->post_title;
+				}
+
+				$buyer = $buyer_map[ $access_token ];
```

Comment:
The memoization doesn't hurt anything, but I wonder if it's addressing a symptom rather than the root cause? IIRC, all of this code is pretty old, so it doesn't seem like it would be the cause of the recent problems. I wonder what's changed between now and when this was performant.

I don't think it's the DC migration, since that should have improved speed. I don't see anything obvious in the commit history, but haven't dug too deep or too far back.

Reply:
So without memoization, this code is doing a  `n+1` query pattern, so I would say we definitely need to fix this. Ideally I would want to use `WP_Query` as you suggest in an above comment, but that means I would need to spend a bit more time. 
I guess it was non that performant earlier,  but its just that this may be the first event where request time limit is getting hit?

Reply:
I think that post meta is normally fetched and cached when the post itself is fetched, so `get_post_meta()` should be fetching from memory instead of making a db query. If that's the case, then it doesn't seem like reducing the # of fetches from memory will make a noticeable difference.

IIRC, @coreymckrill was saying that he's seen this bug on smaller sites as well, and WCEU 2018 didn't run into it.

Reply:
Also, I tried timing different different part of the loop, and this was the part taking significantly more time than the other ones.

Reply:
Oh so this memoization is for `get_posts` call that was earlier on line 2817 and not a `get_post_meta` call. Did you meant that `get_posts` call is also fetching from memory instead from db?

Reply:
Yeah, I'm not opposed to merging this, but it seems like it might be worth digging a bit deeper first, since it sounds like we'll still need some other fixes to really make the code performant again. It might turn out that there's something that has a much bigger impact than this.

How long does the a site with ~500 attendees take before and after applying this? If WCUS is still taking 30 seconds, it seems like there's still a big problem. I would think it'd only take roughly 5 or less.

Reply:
D'oh, you're right, I'd missed that you were replacing that `get_posts()` call w/ just a  `get_post_meta()`. That makes more sense now :)

It seems like there may still be a bigger problem, but we'll probably want to keep this memoization either way.

Reply:
I did find some other cases, (see below comment) but turns out this was indeed the biggest win.